### PR TITLE
adapter: Fix timestamp selection of constant materialized views

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, a materialized view with a constant query without temporal filters was considered timestamp independent, as in the adapter would be free to choose any timestamp for the query. While it's true that the timestamp selected would not affect the answer to the query, it would affect our ability to query the materialized view. Materialized views must durably materialize the results of their definition at a given timestamp before we can query this timestamp. This commit updates the timeline context logic to accurately reflect this.

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fixes timestamp selection for constant materialized views.

### CODE REVIEW

File: src/adapter/src/coord/timeline.rs [hunk 1]
```diff
@@ -521,6 +520,16 @@ impl Coordinator {
                         }
                         ids.extend(optimized_expr.depends_on());
                     }
+                    CatalogItem::MaterializedView(MaterializedView { optimized_expr, .. }) => {
+                        // In some cases the timestamp selected may not affect the answer to a
+                        // query, but it may affect our ability to query the materialized view.
+                        // Materialized views must durably materialize the result of a query, even
+                        // for constant queries. If we choose a timestamps larger than the upper,
```
Comment (umanwizard) [inline src/adapter/src/coord/timeline.rs:527]:
```suggestion
                        // for constant queries. If we choose a timestamp larger than the upper,
```

**Review by umanwizard — APPROVED** [review]
I found it hard to evaluate whether this change is correct, because `get_timeline_contexts` and `TimelineContext` both have circular doc comments; e.g., the one for `TimelineContext` is just "An enum describing the timeline context of a query". They don't really explain what a "timeline context" is, what it means, or how it influences query timestamp selection. It would be good if you could document this a bit more as part of landing your change.

**Comment by jkosh44** [comment]
> I found it hard to evaluate whether this change is correct, because `get_timeline_contexts` and `TimelineContext` both have circular doc comments; e.g., the one for `TimelineContext` is just "An enum describing the timeline context of a query". They don't really explain what a "timeline context" is, what it means, or how it influences query timestamp selection. It would be good if you could document this a bit more as part of landing your change.

I updated the docs, let me know if that clears things up. `TimelineContext` affects what timeline is used during query timestamp selection and whether or not a timestamp is selected at all. The comments over the individual variants are also pretty useful.

**Review by umanwizard — APPROVED** [review]
The change seems legit from our perspective. We should just accept whatever range of validity the MV tells us.

From a compute perspective, though, I'm not sure why an MV on a constant collection doesn't just report all of time for its validity range.

**Comment by jkosh44** [comment]
> From a compute perspective, though, I'm not sure why an MV on a constant collection doesn't just report all of time for its validity range.

The MV must first durably write down the results at a given timestamp `ts` before reporting that it's upper has advanced past `ts`. A constant collection will eventually advance it's upper to `[]` and report all of time for it's validity range, but if for some reason that process is slow or stuck, then we may want to use the current upper during timestamp selection. The motivating example for this was a constant MV created on a cluster with 0 replicas. It's upper will be stuck at `0` and it will not report all of time for it's validity range.

**Pull request merged**